[ESP32] Platform diagnostics framework

pimpalemahesh opened the pull request:
### Problem
The existing implementation lacks the capability to collect and retrieve diagnostics using the diagnostic-logs cluster.

### Change Overview

1. Platform Diagnostic Framework:
Introduced a framework to collect, store, and retrieve diagnostic data using a TLVCircularBuffer, ensuring efficient memory usage and standardized TLV formatting.

2. Backend Implementation
Integrated diagnostic data collection through Matter Traces, with all logs stored in the TLVCircularBuffer.

3. Application Integration
Integrated the diagnostics framework into the esp32-temperature-measurement example application.

### Testing
**Unit Testing**

Added unit tests for:

- Encoding and decoding diagnostic log entries using TLV
- Storing and retrieving entries from the circular buffer
- Retrieving logs multiple times even when buffer size is limited

All tests passed successfully with the current changes.

**Manual Testing**

- Target Platform: ESP32-C3
- Build Configuration: Enabled `ENABLE_ESP_DIAGNOSTICS_TRACE` during the build of the esp32-temperature-measurement app.
- Commissioning and Retrieval:
  Device successfully commissioned using chip-tool.
  Logs confirmed to be written to diagnostic storage.
  Retrieved logs using the Diagnostic Logs cluster over BDX.

- Command Used:
  chip-tool diagnosticlogs retrieve-logs-request 0 1 1 0 --TransferFileDesignator user5.log

Device Logs:
```
I (24280) app-devicecallbacks: Commissioning complete
I (38340) chip[DL]: Total Retrieved bytes: 3036 (82 entries)
```

Observed Output (Chip Tool):
```
[1744176320.069] [49422:49424] [ATM] Handling received BDX Message
[1744176320.069] [49422:49424] [ATM] SendInit
[1744176320.069] [49422:49424] [ATM]   Proposed Transfer Control: 0x10
[1744176320.069] [49422:49424] [ATM]   Range Control: 0x0
[1744176320.069] [49422:49424] [ATM]   Proposed Max Block Size: 1024
[1744176320.069] [49422:49424] [ATM]   Start Offset: 0x0000000000000000
[1744176320.069] [49422:49424] [ATM]   Proposed Max Length: 0x0000000000000000
[1744176320.069] [49422:49424] [ATM]   File Designator Length: 9
[1744176320.069] [49422:49424] [ATM]   File Designator: user5.log
[1744176320.119] [49422:49424] [BDX] Got an event InitReceived
[1744176320.119] [49422:49424] [TOO] OnTransferBegin (9): user5.log
[1744176320.119] [49422:49424] [ATM] Sending BDX Message
[1744176320.119] [49422:49424] [ATM] SendAccept
[1744176320.120] [49422:49424] [ATM]   Transfer Control: 0x10
[1744176320.120] [49422:49424] [ATM]   Max Block Size: 1024
[1744176320.585] [49422:49424] [ATM] Handling received BDX Message
[1744176320.585] [49422:49424] [ATM] Block
[1744176320.585] [49422:49424] [ATM]   Block Counter: 0
[1744176320.585] [49422:49424] [ATM]   Data Length: 1024
[1744176320.621] [49422:49424] [TOO] OnTransferData (9): user5.log
[1744176320.621] [49422:49424] [ATM] Sending BDX Message
[1744176320.621] [49422:49424] [ATM] BlockAck
[1744176320.621] [49422:49424] [ATM]   Block Counter: 0
[1744176321.078] [49422:49424] [ATM] Handling received BDX Message
[1744176321.078] [49422:49424] [ATM] Block
[1744176321.078] [49422:49424] [ATM]   Block Counter: 1
[1744176321.078] [49422:49424] [ATM]   Data Length: 1024
[1744176321.123] [49422:49424] [BDX] Got an event BlockReceived
[1744176321.123] [49422:49424] [TOO] OnTransferData (9): user5.log
[1744176321.123] [49422:49424] [ATM] Sending BDX Message
[1744176321.123] [49422:49424] [ATM] BlockAck
[1744176321.123] [49422:49424] [ATM]   Block Counter: 1
[1744176321.486] [49422:49424] [ATM] Handling received BDX Message
[1744176321.486] [49422:49424] [ATM] BlockEOF
[1744176321.486] [49422:49424] [ATM]   Block Counter: 2
[1744176321.486] [49422:49424] [ATM]   Data Length: 988
[1744176321.526] [49422:49424] [BDX] Got an event BlockReceived
[1744176321.526] [49422:49424] [TOO] OnTransferData (9): user5.log
[1744176321.526] [49422:49424] [ATM] Sending BDX Message
[1744176321.526] [49422:49424] [ATM] BlockAckEOF
[1744176321.526] [49422:49424] [ATM]   Block Counter: 2
```

**Parsed Output sample:**
```
Timestamp | Label | Value
-- | -- | --
15100 | OnMessageReceived | PASESession
15100 | HandlePBKDFParamRequest | PASESession
16130 | Pake3 | 1
```

pimpalemahesh committed:
esp32 diagnostic trace

- Working backend with metric, trace and counter diagnostics
- Diagnostic interface implementation with ring buffer storage
- Added option ENABLE_ESP_DIAGNOSTICS_TRACE in chip KConfig
- Added required options for enabling matter diagnostic trace in project Kconfig

pimpalemahesh committed:
example/temperature-measurement-app
- Resolve buffer issues
- Use single buffer for store and retrieve of diagnostics
- Resolve data loss issue

pimpalemahesh committed:
backend: Add description for diagnosticstorage interface, remove unncessary comments, format files, namespace changes

bzbarsky-apple commented:
Would have been helpful if something in the PR title or description made it clear this is esp32-only...

shubhamdp commented on src/tracing/esp32_diagnostic_trace/Counter.cpp:
```diff
@@ -0,0 +1,70 @@
+    void * ptr = Platform::MemoryAlloc(sizeof(ESPDiagnosticCounter));
+    VerifyOrDie(ptr != nullptr);
+
+    ESPDiagnosticCounter * newInstance = new (ptr) ESPDiagnosticCounter(label);
```
I would suggest using `chip::Platform::MemoryAlloc`, new will throw the exception if it fails to allocate the memory, and we are not handling it.

pimpalemahesh committed:
temperature_measurement_app: Review changes

pimpalemahesh committed:
temperature_measurement_app: Review changes

pimpalemahesh committed:
backend: map for counter diagnostics related changes
- esp32_diagnostic_trace: improve storage class design
- add extra tlv closing bytes check before copying diagnostic
- unify diagnostic entries into a single type

pimpalemahesh committed:
esp32_diagnostic_trace: Review changes
- backend: Replace linkedlist with map for counter diagnostics
- backend: Pass diagnostic storage as a parameter to backend
- esp32_diagnostic_trace: Return actual data size

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=36532) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=36532) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: pimpalemahesh<br/>:x: esp<br/><hr/>**esp** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=36532) it.</sub>

pimpalemahesh committed:
example: Remove redundant code

pimpalemahesh committed:
diagnostic-buffer: clear buffer after successful bdx log transfer
- Add related logic in temperature-measurement-app

shubhamdp commented on src/tracing/esp32_diagnostic_trace/Diagnostics.h:
```diff
@@ -0,0 +1,127 @@
+        chip::TLV::TLVType DiagnosticOuterContainer = chip::TLV::kTLVType_NotSpecified;
+        ReturnErrorOnFailure(
+            writer.StartContainer(chip::TLV::AnonymousTag(), chip::TLV::kTLVType_Structure, DiagnosticOuterContainer));
+        ReturnErrorOnFailure(writer.Put(chip::TLV::ContextTag(0), timestamp_));
```
we should define these tag fields with some constants, it would be easier to understand and the same enums can be used in the python script that is utilizing this data.

pimpalemahesh committed:
diagnostic-storage: Remove singltone design for storage instance
- Add private CircularTLVReader and CircularTLVWriter
- Remove redundant outer container for tlv's from Retrieve method
- Maintain TAG's for diagnostic entry elements
- Make diagnostic entry as a constant param to store method
- Move Murmurhash to utils namespace

pimpalemahesh committed:
temperature-measurement-app: use seperate buffer for retrieval of diagnostics
- move diagnostic storage buffer to main

shubhamdp commented on src/tracing/esp32_diagnostic_trace/DiagnosticStorageManager.h:
```diff
@@ -0,0 +1,115 @@
+        return CHIP_NO_ERROR;
+    }
+
+    bool IsEmptyBuffer() override { return DataLength() == 0; }
```
```suggestion
    bool IsBufferEmpty() override { return DataLength() == 0; }
```

shubhamdp commented on src/tracing/esp32_diagnostic_trace/DiagnosticTracing.cpp:
```diff
@@ -0,0 +1,169 @@
+    switch (event.ValueType())
+    {
+    case ValueType::kInt32: {
+        ESP_LOGI("mtr", "The value of %s is %ld ", event.key(), event.ValueInt32());
```
```suggestion
        ESP_LOGI("mtr", "The value of %s is %" PRId32, event.key(), event.ValueInt32());
```

I'd suggest using the width specific format specifiers, and it would be good to use `ChipLogProgress` instead of ESP_LOGx

shubhamdp commented on src/tracing/esp32_diagnostic_trace/DiagnosticTracing.cpp:
```diff
@@ -0,0 +1,169 @@
+    }
+}
+
+void ESP32Diagnostics::StoreDiagnostics(const char * label, const char * group)
```
should we propagate the error a level up?

shubhamdp commented on src/tracing/esp32_diagnostic_trace/DiagnosticTracing.cpp:
```diff
@@ -0,0 +1,169 @@
+        break;
+    }
+
+    VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(DeviceLayer, "Failed to store Metric Diagnostic data"));
```
should we propagate the error a level up?

pimpalemahesh replied:
Here it is not possible as this method is overridden from backend.

shubhamdp commented on src/tracing/esp32_diagnostic_trace/Counter.cpp:
```diff
@@ -0,0 +1,55 @@
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    Diagnostic<uint32_t> counter(label, GetInstanceCount(label), esp_log_timestamp());
+    err = mStorageInstance.Store(counter);
```
should we propagate the error to level above?

shubhamdp commented on src/tracing/esp32_diagnostic_trace/Diagnostics.h:
```diff
@@ -0,0 +1,135 @@
+namespace Diagnostics {
+
+// Diagnostic TAGs
+enum D_TAG
+{
+    TIMESTAMP = 0,
+    LABEL,
+    VALUE
+};
```
```suggestion
enum class DiagTag : uint8_t
{
    kTimestamp = 0,
    kLabel,
    kValue,
};
```

just trying to align with code-style, and then the usage will be `DiagTag::kValue`, it makes reading the code easier as well.

pimpalemahesh committed:
temperature-measurement-app: pass diagnostic storage Instance as pointer to LogProvider
- add kconfig option for retrieval buffer size

pimpalemahesh committed:
backend: update methods to propagate error to level up
- pass diagnostic storage instance as a pointer to backend
- replace ESPLog statement with ChipLog for logging
diagnostic-storage: update IsEmptyBuffer method to IsBufferEmpty
- change diagnostic TAG enum class

shubhamdp commented on examples/temperature-measurement-app/esp32/main/diagnostic-logs-provider-delegate-impl.cpp:
```diff
@@ -29,6 +29,11 @@ using namespace chip::app::Clusters::DiagnosticLogs;
 LogProvider LogProvider::sInstance;
 LogProvider::CrashLogContext LogProvider::sCrashLogContext;
 
+#if CONFIG_ENABLE_ESP_DIAGNOSTICS_TRACE
```
```suggestion
#ifdef CONFIG_ENABLE_ESP_DIAGNOSTICS_TRACE
```
else `-Wundef ` would throw error.

Please change the other usage of this config as well

shubhamdp commented on examples/temperature-measurement-app/esp32/main/diagnostic-logs-provider-delegate-impl.cpp:
```diff
@@ -108,8 +121,24 @@ CHIP_ERROR LogProvider::PrepareLogContextForIntent(LogContext * context, IntentE
+        CHIP_ERROR err = mStorageInstance->Retrieve(endUserSupportSpan, sReadEntries);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(DeviceLayer, "Failed to retrieve data: %s", chip::ErrorStr(err));
```
```suggestion
            ChipLogError(DeviceLayer, "Failed to retrieve data: %" CHIP_ERROR_FORMAT, err.Format());
```

Please check for other usage of `ErrorStr` and replace it with `.Format()`

shubhamdp commented on examples/temperature-measurement-app/esp32/main/diagnostic-logs-provider-delegate-impl.cpp:
```diff
@@ -278,11 +307,23 @@ CHIP_ERROR LogProvider::StartLogCollection(IntentEnum intent, LogSessionHandle &
+        }
+        else
+        {
+            ChipLogProgress(DeviceLayer, "Cleared all read diagnostics successfully");
```
This will be printed after every call to `EndLogCollection()`, if this is too much of logs, may be can we make it `ChipLogDetail()`?

shubhamdp commented on examples/temperature-measurement-app/esp32/main/include/diagnostic-logs-provider-delegate-impl.h:
```diff
@@ -39,7 +45,16 @@ namespace DiagnosticLogs {
 {
 public:
+#ifdef CONFIG_ENABLE_ESP_DIAGNOSTICS_TRACE
+    static inline LogProvider & GetInstance(CircularDiagnosticBuffer * bufferInstance)
```
Either there should be a setter API for setting bufferInstance or may be a constructor override?

shubhamdp replied:
we should document the behaviour in case of `nullptr`.

shubhamdp commented on examples/temperature-measurement-app/esp32/main/main.cpp:
```diff
@@ -75,14 +82,18 @@ static AppDeviceCallbacks EchoCallbacks;
 {
     Esp32AppServer::Init(); // Init ZCL Data Model and CHIP App Server AND Initialize device attestation config
+#if CONFIG_ENABLE_ESP_DIAGNOSTICS_TRACE
+    diagnosticStorage.Init(endUserBuffer, CONFIG_END_USER_BUFFER_SIZE);
```
why are we passing `endUserBuffer` twice? first time in the constructor and second time here?

shubhamdp commented on src/tracing/esp32_diagnostic_trace/Counter.cpp:
```diff
@@ -0,0 +1,56 @@
+    return mCounterList[label];
+}
+
+CHIP_ERROR ESPDiagnosticCounter::ReportMetrics(const char * label, DiagnosticStorageInterface * mStorageInstance)
```
```suggestion
CHIP_ERROR ESPDiagnosticCounter::ReportMetrics(const char * label, DiagnosticStorageInterface * storageInstance)
```

variable starting with `m` says that they are member of that class, so not needed for arg or local variables

shubhamdp commented on src/tracing/esp32_diagnostic_trace/Counter.cpp:
```diff
@@ -0,0 +1,56 @@
+CHIP_ERROR ESPDiagnosticCounter::ReportMetrics(const char * label, DiagnosticStorageInterface * mStorageInstance)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrReturnError(mStorageInstance != nullptr, err, ChipLogError(DeviceLayer, "Diagnostic Storage Instance cannot be NULL"));
```
should it return some error here, instead of CHIP_NO_ERROR?

shubhamdp replied:
We should return INCORRECT_STATE.

shubhamdp commented on src/tracing/esp32_diagnostic_trace/Counter.cpp:
```diff
@@ -0,0 +1,56 @@
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrReturnError(mStorageInstance != nullptr, err, ChipLogError(DeviceLayer, "Diagnostic Storage Instance cannot be NULL"));
+    Diagnostic<uint32_t> counter(label, GetInstanceCount(label), esp_log_timestamp());
+    err = mStorageInstance->Store(counter);
+    return err;
```
```suggestion
    return mStorageInstance->Store(counter);
```

and we do not need that declaration on line 47.

shubhamdp commented on src/tracing/esp32_diagnostic_trace/DiagnosticTracing.cpp:
```diff
@@ -0,0 +1,175 @@
+
+void ESP32Diagnostics::TraceBegin(const char * label, const char * group)
+{
+    if (IsPermitted(group))
+    {
+        CHIP_ERROR err = StoreDiagnostics(label, group);
+        VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(DeviceLayer, "Failed to store Trace Diagnostic data"));
+    }
```
```suggestion
    VerifyOrReturn(IsPermitted(group));
    ReturnOnFailure(StoreDiagnostics(label, group));
```

pimpalemahesh committed:
temperature-measurement-app: review changes

pimpalemahesh committed:
esp32_diagnostic_trace: review changes

pimpalemahesh committed:
esp32_diagnostic_trace: Add clearBuffer() method, update api documentation

pimpalemahesh committed:
esp_diagnostic_trace: add clearbuffer method to diagnostic interface

pimpalemahesh committed:
fixed trace instant filter

pimpalemahesh committed:
diagnostics: add decode method

shubhamdp commented on src/tracing/esp32_diagnostic_trace/Diagnostics.h:
```diff
@@ -0,0 +1,249 @@
+class Diagnostic : public DiagnosticEntry
+{
+public:
+    Diagnostic(char * label = nullptr, T value = T{}, uint32_t timestamp = 0) : label_(label), value_(value), timestamp_(timestamp)
```
```suggestion
    Diagnostic(const char * label = nullptr, T value = T{}, uint32_t timestamp = 0) : label_(label), value_(value), timestamp_(timestamp)
```

shubhamdp commented on src/tracing/esp32_diagnostic_trace/Diagnostics.h:
```diff
@@ -0,0 +1,249 @@
+        }
+        else
+        {
+            ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::LABEL), label_));
```
may be we can use `PutString()`'s this variant which takes length as input, and we don't need an additional buffer, and length is enough? https://github.com/project-chip/connectedhomeip/blob/master/src/lib/core/TLVWriter.h#L585

```suggestion
            ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::LABEL), label_), std::min(strlen(label_), kMaxStringValueSize));
```

shubhamdp commented on src/tracing/esp32_diagnostic_trace/Diagnostics.h:
```diff
@@ -0,0 +1,249 @@
+        }
+        ReturnErrorOnFailure(writer.EndContainer(DiagnosticOuterContainer));
+        ReturnErrorOnFailure(writer.Finalize());
+        ChipLogProgress(DeviceLayer, "Diagnostic Value written to storage successfully. label: %s\n", label_);
```
```suggestion
        ChipLogDetail(DeviceLayer, "Diagnostic Value written to storage successfully. label: %s\n", label_);
```

I think this may generate too much logs, so lets make this a debug print?

shubhamdp commented on src/tracing/esp32_diagnostic_trace/Diagnostics.h:
```diff
@@ -0,0 +1,249 @@
+    uint32_t GetTimestamp() const { return timestamp_; }
+
+private:
+    char * label_ = nullptr;
+    T value_{};
+    uint32_t timestamp_ = 0;
```
```suggestion
    const char * mLabel = nullptr;
    T mValue{};
    uint32_t mTimestamp = 0;
```

just to be more consistent with overall code base

shubhamdp commented on src/tracing/esp32_diagnostic_trace/DiagnosticTracing.h:
```diff
@@ -0,0 +1,66 @@
+namespace Diagnostics {
+/// A Backend that outputs data to chip logging.
+///
+/// Structured data is formatted as json strings.
```
are we not using tlv?

shubhamdp commented on src/tracing/esp32_diagnostic_trace/DiagnosticTracing.cpp:
```diff
@@ -0,0 +1,170 @@
+    break;
+
+    case ValueType::kChipErrorCode:
+        ChipLogProgress(DeviceLayer, "The value of %s is error with code %lu ", event.key(), event.ValueErrorCode());
```
Do we need these logs? can we change them to Detail?

shubhamdp commented on src/tracing/esp32_diagnostic_trace/DiagnosticTracing.cpp:
```diff
@@ -0,0 +1,170 @@
+
+// All traces with value from gSkipList are skipped.
+// Used for MATTER_TRACE_INSTANT()
+HashValue gSkipList[kPermitListMaxSize] = {
```
can you add a comment in code why are we skipping few tags?

shubhamdp commented on src/tracing/esp32_diagnostic_trace/DiagnosticStorageManager.h:
```diff
@@ -0,0 +1,111 @@
+            successful_written_bytes = writer.GetLengthWritten();
+        }
+        span.reduce_size(successful_written_bytes);
+        ChipLogProgress(DeviceLayer, "---------------Total Retrieved bytes : %ld----------------\n", successful_written_bytes);
```
please change this to Detail, not just these but the logs that are added for debugging purpose and would bloat the output should be changed to Detail.

pimpalemahesh committed:
readme update

pimpalemahesh committed:
Review changes

pimpalemahesh committed:
Modified chip log statements, make string values mutable

pimpalemahesh committed:
Restyled changes

pimpalemahesh committed:
backend: copyright header change, retrieve api return value changed

pimpalemahesh committed:
diagnosticEntry: add invalid valuetype

andy31415 commented on src/tracing/esp32_diagnostic_trace/Counter.cpp:
```diff
@@ -0,0 +1,60 @@
+namespace Tracing {
+namespace Diagnostics {
+
+std::map<const char *, uint32_t> ESPDiagnosticCounter::mCounterList;
```
these compares will not be `strcmp` like compares but rather memory pointer compares. Is this expected/intended?

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticTracing.cpp:
```diff
@@ -0,0 +1,187 @@
+namespace Tracing {
+namespace Diagnostics {
+
+namespace Utils {
```
how about a `Hashing` namespace instead utils? utils/support/core/helpers generally have no meaning.

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticTracing.cpp:
```diff
@@ -0,0 +1,187 @@
+
+constexpr size_t kPermitListMaxSize = CONFIG_MAX_PERMIT_LIST_SIZE;
+using HashValue                     = uint32_t;
+using namespace Utils;
```
why declare a namespace to use it right away? Anonymous namespace without using seems better.

pimpalemahesh committed:
Review comments

andy31415 commented on src/tracing/esp32_diagnostic_trace/Counter.h:
```diff
@@ -0,0 +1,64 @@
+class ESPDiagnosticCounter
+{
+public:
+    static ESPDiagnosticCounter & GetInstance(const char * label)
```
This API needs documentation: why does a `Get` have sideffects (it increases a label count)?

I would rather expect to call `GetInstance()->IncreaseCount(label)` in code. Why is the instance returned? All functions that are public take in a label argument ... so even `GetInstance("foo").ReportMetrics("foo", storage)` does not seem to make sense.

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticEntry.cpp:
```diff
@@ -0,0 +1,137 @@
+    // Write label
+    if (entry.label != nullptr)
+    {
+        uint32_t labelSize = strlen(entry.label) > kMaxStringValueSize ? kMaxStringValueSize : strlen(entry.label);
+        if (labelSize >= kMaxStringValueSize)
+        {
+            char labelBuffer[kMaxStringValueSize + 1];
+            memcpy(labelBuffer, entry.label, kMaxStringValueSize);
+            labelBuffer[kMaxStringValueSize] = '\0';
+            ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::LABEL), labelBuffer));
+        }
+        else
+        {
+            ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::LABEL), entry.label, labelSize));
+        }
```
```suggestion
        char labelBuffer[kMaxStringValueSize + 1];
        Platform::CopyString(labelBuffer, entry.label);
        ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::LABEL), labelBuffer));
```

Seems shorter, although less performant. If you want the performance optimization to elide copy that is ok as well, however please use CopyString (from `support/CHIPMemString.h`) instead of implementing your own safe strcpy.

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticEntry.cpp:
```diff
@@ -0,0 +1,137 @@
+    case ValueType::kCharString:
+        if (entry.stringValue != nullptr)
+        {
+            uint32_t valueSize = strlen(entry.stringValue) > kMaxStringValueSize ? kMaxStringValueSize : strlen(entry.stringValue);
```
same comment here: you can make this sorter or at least use CopyString.

This seems like duplicated code ... can we reuse?

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticEntry.h:
```diff
@@ -0,0 +1,63 @@
+
+struct DiagnosticEntry
+{
+    // mutable label because modified in decode
```
please explain the lifetime of this. We seem to do TLVReader::Get(, kMaxStringValueSize + 1) for both label and string value to read data. Who allocates the data and for how long? since the value is a union, there is no destructor cleanup and value is owned by someone else. Pleas explain usage and the size requirement (which seems very important)

pimpalemahesh replied:
There is no specific requirement for external memory allocation. I will use fixed arrays instead, which will reduce the overhead of memory management and tie the memory lifetime directly to the struct itself. Thanks.

andy31415 reviewed (changes requested):
Several files in esp32_diagnostic_trace are stand-alone and do not depend on 3rd party integration with diagnostics directly it seems. please add unit tests for them (for example I that encode/decode works correctly).

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticEntry.cpp:
```diff
@@ -0,0 +1,137 @@
+    TLVType containerType;
+    ReturnErrorOnFailure(reader.EnterContainer(containerType));
+
+    // Read timestamp
```
these comments just re-state the code. Do we need to write anything? I would suggest removing or updating so that they say things that are not obvious from the code.

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticEntry.h:
```diff
@@ -0,0 +1,63 @@
+        int32_t intValue;
+    };
+    ValueType type;
+    uint32_t timestamp;
```
what is the `timestamp` represented it? Is it epoch? Seconds since boot? Milliseconds since boot? Something else?

Please add a suffix both here and in the enum to explain, like `timestamps_sec_since_boot` or `timestamp_epoch_s`. Also explain in a comment why 32-bit is sufficient (i.e. for how long it seems valid / until what year)

pimpalemahesh replied:
The api I am using to get timestamp returns uint32_t value always so this is sufficient I think.

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticEntry.cpp:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    // Write value based on type
+    switch (entry.type)
```
where is entry.type read? How do we know it is a valid value and corresponds to the TLV value we need?

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticEntry.cpp:
```diff
@@ -0,0 +1,137 @@
+
+    // Read value
+    ReturnErrorOnFailure(reader.Next());
+    switch (entry.type)
```
Where is entry.type read? How do we know it is correct for the TLV data set?
It seems we should either read it or switch on the TLV type and set it instead.

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticEntry.cpp:
```diff
@@ -0,0 +1,137 @@
+        break;
+    }
+
+    ReturnErrorOnFailure(reader.ExitContainer(containerType));
+    return CHIP_NO_ERROR;
```
```suggestion
    return reader.ExitContainer(containerType);
```

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticStorage.cpp:
```diff
@@ -0,0 +1,106 @@
+    bool close_success                = true; // To check if the last TLV is copied successfully.
+    uint32_t successful_written_bytes = 0;    // Store temporary writer length in case last TLV is not copied successfully.
+
+    while ((err = mReader.Next()) == CHIP_NO_ERROR)
```
do we need to handle CHIP_END_OF_TLV differently than others (i.e. end of TLV is essentially success read all, where as other documented errors like TLV_UNDERRUN or INVALID_TLV_ELEMENT are log-worthy failures)

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticStorage.cpp:
```diff
@@ -0,0 +1,106 @@
+        }
+        else
+        {
+            ChipLogDetail(DeviceLayer, "Skipping unexpected TLV element");
```
consider logging the type/tag for easier debugging.

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticStorage.cpp:
```diff
@@ -0,0 +1,106 @@
+    writer.Init(span.data(), span.size());
+    read_entries = 0;
+
+    bool close_success                = true; // To check if the last TLV is copied successfully.
```
given the comment and usage, maybe a better name is `element_copy_success` or similar.
Instead of having this variable, why not use `err` variable instead?

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticStorage.cpp:
```diff
@@ -0,0 +1,106 @@
+    ReturnErrorOnFailure(writer.Finalize());
+    if (close_success)
+    {
+        successful_written_bytes = writer.GetLengthWritten();
```
why is this needed? don't we set it in the loop after every successful copy?

andy31415 replied:
If finalize guarantees no size changes, then this is redundant. If finalize does size changes, then it looks like we miss those changes in case of failures (but still have the copy size).

andy31415 commented on src/tracing/esp32_diagnostic_trace/DiagnosticStorage.cpp:
```diff
@@ -0,0 +1,106 @@
+            ChipLogDetail(DeviceLayer, "Skipping unexpected TLV element");
+        }
+    }
+    ReturnErrorOnFailure(writer.Finalize());
```
if copy fails, is it safe to Finalize? Why do we finalize without updating successfull written bytes?

It seems to me we need some snapshot capability instead: try to copy and if you fail to copy, revert to the snapshot to undo anything the copy tried to do and failed. Copy may have copied partial data.

andy31415 commented:
@pimpalemahesh please add unit tests where you can. Beyond that we are strongly trying to encourage people to write unit tests, so we make it less convenient to write essentially `manually tested` in descriptions. This means `Verified commissioning and retrieval of diagnostics using the diagnostic-logs cluster over BDX on the ESP32-C3 chip with the temperature-measurement-app example.` is insufficient. Please add:
   - what application you built (e.g. what build-examples target)
   - what chip-tool commands you used to verify and what were the observed outputs (not full logs, those are too big, but the extracted relevant lines that show correct behavior)
   - explanation of why sections of this CANNOT be tested automatically.

pimpalemahesh committed:
review comments
-DiagnosticEntry structure change
-Encode, Decode, Retrieve diagnostic method changes

pimpalemahesh committed:
fix read entryes type in log prints

pimpalemahesh committed:
esp29_diagnostic_trace: Add tests for Encode, Decode, Store and Retrieve methods of the diagnostic storage

pimpalemahesh commented:
> @pimpalemahesh please add unit tests where you can. Beyond that we are strongly trying to encourage people to write unit tests, so we make it less convenient to write essentially `manually tested` in descriptions. This means `Verified commissioning and retrieval of diagnostics using the diagnostic-logs cluster over BDX on the ESP32-C3 chip with the temperature-measurement-app example.` is insufficient. Please add:
> 
>     * what application you built (e.g. what build-examples target - in your case it seems a esp32-temperature-measurement variant)
> 
>     * what chip-tool commands you used to verify and what were the observed outputs (not full logs, those are too big, but the extracted relevant lines that show correct behavior)
> 
>     * explanation of why sections of this CANNOT be tested automatically.

Added Unit Tests and updated PR Description as per the suggestion

andy31415 commented on src/BUILD.gn:
```diff
@@ -124,6 +124,12 @@ if (chip_build_tests) {
 
+    if (matter_enable_tracing_support &&
+        matter_trace_config ==
+        "${chip_root}/src/tracing/esp32_diagnostic_trace") {
```
Is this needed? testable code should be generic. Can we split things into generic code (that likely does not need esp32-specific includes and then can be tested indepently) and esp-specific code (which likely has no unit tests)?

pimpalemahesh committed:
review comments
- moved diagnostic storage specific tests to the esp32_diagnostic_trace
  folder
- removed dependency of tracing and esp32 specific code form tests

pimpalemahesh committed:
replace size_t with uint32_t for diagnostic storage

pimpalemahesh committed:
diangostic test: converted int value to uint to address implicit convertsion error
